src/Castle.Core.AsyncInterceptor/InterceptorAdapter.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) 2016 James Skimming. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See LICENSE in the project root for license information.
+
+namespace Castle.DynamicProxy
+{
+    using System.Reflection;
+    using System.Runtime.CompilerServices;
+    using System.Runtime.ExceptionServices;
+    using System.Threading.Tasks;
+
+    using Castle.DynamicProxy.Invocations;
+
+    /// <inheritdoc cref="IInterceptor"/>
+    internal class InterceptorAdapter : IInterceptor
+    {
+        private static readonly MethodInfo InterceptAsyncFunctionMethodInfo = typeof(InterceptorAdapter).GetMethod(
+            nameof(ExecuteForAsyncFunction),
+            BindingFlags.Instance | BindingFlags.NonPublic);
+
+        private static readonly MethodInfo InterceptFunctionMethodInfo = typeof(InterceptorAdapter).GetMethod(
+            nameof(ExecuteForFunction),
+            BindingFlags.Instance | BindingFlags.NonPublic);
+
+        private readonly IAsyncInterceptor _interceptor;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="InterceptorAdapter"/> class.
+        /// </summary>
+        /// <param name="interceptor">
+        /// The interceptors.
+        /// </param>
+        public InterceptorAdapter(IAsyncInterceptor interceptor)
+        {
+            _interceptor = interceptor;
+        }
+
+        /// <summary>
+        /// Intercepts a method <paramref name="invocation"/>.
+        /// </summary>
+        /// <param name="invocation">The method invocation.</param>
+        public void Intercept(IInvocation invocation)
+        {
+            try
+            {
+                if (invocation.Method.ReturnType == typeof(void))
+                {
+                    _interceptor.InterceptAction(new ActionInvocation(invocation));
+                }
+                else if (invocation.Method.ReturnType == typeof(Task))
+                {
+                    invocation.ReturnValue = _interceptor.InterceptAsyncAction(new AsyncActionInvocation(invocation));
+                }
+                else if (typeof(Task).IsAssignableFrom(invocation.Method.ReturnType))
+                {
+                    invocation.ReturnValue = InterceptAsyncFunctionMethodInfo
```

Comment:
Is there any reason to remove the "concurrent dictionary" implementation based on method delegates?

Reply:
No, I was under the assumption that performance cost was Getting the `MethodInfo` and not calling Invoke. After doing a few benchmarks it was clear that the `AsyncDeterminationInterceptor` was superior in performance. 

Reverted back to original implementation with the addition of Function<T> handling